### PR TITLE
attempt to fix MATCH syntax

### PR DESCRIPTION
**Pull request by qiukun**
#### Fix compatibility problem about keyword REMOVE PARTITIONING

Issue: pingcap/parser#426

#### MySQL Syntax:

ReferDef:
	"REFERENCES" TableName MatchOpt OnDeleteUpdateOpt

### Bad SQL Case:

alter table t add column a double (4,2) zerofill references b match full on update set null first;

### Check List

Tests
- Unit test

### CODE REVIEW

**Comment by qiukun** [comment]
The test is failed on FieldOpt part.
In `parser.y:8277`:
```
|	"ZEROFILL"
	{
		$$ = &ast.TypeOpt{IsZerofill: true, IsUnsigned: true}
	}
```
Is this the intended behaviour and I should modify the test or should I remove `IsUnsigned: true`?

**Comment by qiukun** [comment]
https://dev.mysql.com/doc/refman/5.5/en/numeric-type-attributes.html:
`If you specify ZEROFILL for a numeric column, MySQL automatically adds the UNSIGNED attribute.`

I amended the test accordingly.

File: parser_test.go [hunk 1]
```diff
@@ -2297,6 +2297,9 @@ func (s *testParserSuite) TestDDL(c *C) {
 		{"alter table t remove partitioning", true, "ALTER TABLE `t` REMOVE PARTITIONING"},
 		{"alter table db.ident remove partitioning", true, "ALTER TABLE `db`.`ident` REMOVE PARTITIONING"},
 		{"alter table t lock = default remove partitioning", true, "ALTER TABLE `t` LOCK = DEFAULT REMOVE PARTITIONING"},
+
+		// for match
+		{"alter table t add column a double (4,2) zerofill references b match full on update set null first", true, "ALTER TABLE `t` ADD COLUMN `a` DOUBLE(4,2) UNSIGNED ZEROFILL REFERENCES `b` MATCH FULL ON UPDATE SET NULL FIRST"},
```
Comment (leoppro) [inline parser_test.go]: could you add more test cases?

**Comment by leoppro** [comment]
rest LGTM
PTAL @tangenta

File: parser.y [hunk 2]
```diff
@@ -2217,6 +2228,8 @@ ReferDef:
 		}
 	}
 
+
+
```
Comment (tangenta) [inline parser.y]: Please remove these useless lines.

File: parser.y [hunk 3]
```diff
@@ -2205,6 +2205,17 @@ MatchOpt:
...
+		onDeleteUpdate := $4.([2]interface{})
+		$$ = &ast.ReferenceDef{
+			Table: $2.(*ast.TableName),
+			OnDelete: onDeleteUpdate[0].(*ast.OnDeleteOpt),
+			OnUpdate: onDeleteUpdate[1].(*ast.OnUpdateOpt),
+			Match: $3.(ast.MatchType),
+		}
+	}
+|
 	"REFERENCES" TableName '(' IndexColNameList ')' MatchOpt OnDeleteUpdateOpt
```
Comment (tangenta) [inline parser.y]: IMHO, we should abstract `'(' IndexColNameList ')'` as `IndexColNameListOpt` for better reusablility.

**Comment by qiukun** [comment]
More tests are added.
`ASC` is dropped as the current implementation of parser. `ON DELETE` and `ON UPDATE` are adjusted as the mysql doc and the current implementation of parser.

File: parser.y [hunk 4]
```diff
@@ -2205,18 +2206,32 @@ MatchOpt:
...
-		onDeleteUpdate := $7.([2]interface{})
+		onDeleteUpdate := $5.([2]interface{})
+		var indexColNames []*ast.IndexColName
+		if $3 != nil {
+			indexColNames = $3.([]*ast.IndexColName)
+		}
 		$$ = &ast.ReferenceDef{
 			Table: $2.(*ast.TableName),
-			IndexColNames: $4.([]*ast.IndexColName),
+			IndexColNames: indexColNames,
```
Comment (leoppro) [inline parser.y]:
```suggestion
			IndexColNames: $3.([]*ast.IndexColName),
```
and remove L2213-2215

**Comment by leoppro** [comment]
resolve the conflicts please

File: parser_test.go [hunk 5]
```diff
@@ -2297,6 +2297,35 @@ func (s *testParserSuite) TestDDL(c *C) {
...
+		{"alter table d_n.t_n add constraint foreign key ident (ident(1)) references d_n.t_n match full on delete set null", true, "ALTER TABLE `d_n`.`t_n` ADD CONSTRAINT `ident` FOREIGN KEY (`ident`(1)) REFERENCES `d_n`.`t_n` MATCH FULL ON DELETE SET NULL"},
+		{"alter table t_n add constraint ident foreign key (ident,ident(1)) references t_n match full on update set null on delete restrict", true, "ALTER TABLE `t_n` ADD CONSTRAINT `ident` FOREIGN KEY (`ident`, `ident`(1)) REFERENCES `t_n` MATCH FULL ON DELETE RESTRICT ON UPDATE SET NULL"},
+		{"alter table d_n.t_n add foreign key ident (ident, ident(1) asc) references t_n match partial on delete cascade remove partitioning", true, "ALTER TABLE `d_n`.`t_n` ADD CONSTRAINT `ident` FOREIGN KEY (`ident`, `ident`(1)) REFERENCES `t_n` MATCH PARTIAL ON DELETE CASCADE REMOVE PARTITIONING"},
+		{"alter table d_n.t_n add constraint foreign key (ident asc) references d_n.t_n match simple on update cascade on delete cascade", true, "ALTER TABLE `d_n`.`t_n` ADD CONSTRAINT FOREIGN KEY (`ident`) REFERENCES `d_n`.`t_n` MATCH SIMPLE ON DELETE CASCADE ON UPDATE CASCADE"},
+
+		// for character vary syntax #410
```
Comment (leoppro) [inline parser_test.go]:
```suggestion
		// for character vary syntax
```

File: parser.y [hunk 6]
```diff
@@ -8144,10 +8179,27 @@ NationalOpt:
 	{}
```
Comment (leoppro) [inline parser.y]: is `NationalOpt` necessary now?

File: parser_test.go (hunk 5 above)
Comment (qiukun) [reply parser_test.go]: it comes from the master, fixed anyway.

File: parser.y (hunk 6 above)
Comment (qiukun) [reply parser.y]: it comes from the master. I have no idea what it is.

Comment (leoppro) [reply parser.y]: remove it

Comment (leoppro) [reply parser.y]: it was only used in `StringType`, but it's unused now

Comment (leoppro) [reply parser.y]: oh, i am sorry, i mistake this pr into #465, this bug should be fixed in #465, but it has been merged...
could you help me fix it by the way?

Comment (qiukun) [reply parser.y]: My bad, it comes from the auto-merge. Should have check manually. Fixed

File: parser.y [hunk 7]
```diff
@@ -793,6 +793,7 @@ import (
 	IgnoreOptional			"IGNORE or empty"
 	IndexColName			"Index column name"
 	IndexColNameList		"List of index column name"
+	IndexColNameListOpt	    "List of index column name opt"
```
Comment (leoppro) [inline parser.y]:
```suggestion
	IndexColNameListOpt		"List of index column name opt"
```

**Comment by leoppro** [comment]
@tangenta PTAL

File: parser.y [hunk 8]
```diff
@@ -791,8 +791,9 @@ import (
 	IfExists			"If Exists"
 	IfNotExists			"If Not Exists"
 	IgnoreOptional			"IGNORE or empty"
-	IndexColName			"Index column name"
+	IndexColName		"Index column name"
 	IndexColNameList		"List of index column name"
+	IndexColNameListOpt	    "List of index column name opt"
```
Comment (leoppro) [inline parser.y]:
```suggestion
	IndexColNameListOpt		"List of index column name opt"
```

File: parser.y (hunk 7 above)
Comment (qiukun) [reply parser.y]: Indentation is fixed. Now these lines look like their neighborhood.

File: parser.y [hunk 9]
```diff
@@ -791,8 +791,9 @@ import (
 	IfExists			"If Exists"
 	IfNotExists			"If Not Exists"
 	IgnoreOptional			"IGNORE or empty"
-	IndexColName			"Index column name"
+	IndexColName		"Index column name"
```
Comment (leoppro) [inline parser.y]:
```suggestion
	IndexColName			"Index column name"
```

File: parser.y (hunk 8 above)
Comment (qiukun) [reply parser.y]: image
Please recheck the indentation style. I think I did it right.

Comment (leoppro) [reply parser.y]: we usually don't use `space` in yacc file, using `tab`.

**Pull request merged**